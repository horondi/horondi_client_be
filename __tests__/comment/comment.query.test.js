/* eslint-disable no-undef */
const { gql } = require('@apollo/client');
const { COMMENT_NOT_FOUND } = require('../../error-messages/comment.messages');
const { setupApp } = require('../helper-functions');

const {
  newComment,
  commentWrongId,
  validEmail,
  invalidEmail,
  productWrongId,
} = require('./comment.variables');

const {
  createModel,
  newCategory,
  newModel,
  newMaterial,
  getNewProduct,
  deleteAll,
} = require('../product/product.variables');

jest.mock('../../modules/upload/upload.service');
jest.mock('../../modules/currency/currency.model.js');
jest.mock('../../modules/product/product.utils.js');

let commentId = '';
let operations;
let modelId;
let materialId;
let product;
let productId;
let categoryId;
let subcategoryId;

describe('Comment queries', () => {
  beforeAll(async done => {
    operations = await setupApp();
    const itemsId = await createModel(newMaterial, newCategory, newModel);
    modelId = itemsId.modelId;
    materialId = itemsId.materialId;
    categoryId = itemsId.categoryId;
    subcategoryId = itemsId.subcategoryId;

    product = getNewProduct(categoryId, subcategoryId, modelId, materialId);

    const createProduct = await operations.mutate({
      mutation: gql`
        mutation($product: ProductInput!) {
          addProduct(upload: [], product: $product) {
            ... on Product {
              _id
            }
            ... on Error {
              message
            }
          }
        }
      `,
      variables: {
        product,
      },
    });

    productId = createProduct.data.addProduct._id;

    const res = await operations
      .mutate({
        mutation: gql`
          mutation($productId: ID!, $comment: commentInput!) {
            addComment(productId: $productId, comment: $comment) {
              ... on Comment {
                _id
              }
              ... on Error {
                message
                statusCode
              }
            }
          }
        `,
        variables: {
          productId,
          comment: { ...newComment, product: productId },
        },
      })
      .catch(e => e);
    commentId = res.data.addComment._id;
    done();
  });

  afterAll(async done => {
    await deleteAll(materialId, productId, categoryId, modelId);
    await operations
      .mutate({
        mutation: gql`
          mutation($id: ID!) {
            deleteComment(id: $id) {
              ... on Comment {
                _id
              }
              ... on Error {
                statusCode
                message
              }
            }
          }
        `,
        variables: { id: commentId },
      })
      .catch(e => e);
    done();
  });

  it(' Should receive all comments writen by selected user', async done => {
    const res = await operations
      .query({
        variables: {
          userEmail: validEmail,
        },
        query: gql`
          query($userEmail: String!) {
            getAllCommentsByUser(userEmail: $userEmail) {
              ... on Comment {
                text

                product {
                  _id
                }
                show
                user {
                  email
                }
              }
            }
          }
        `,
      })
      .catch(e => e);

    expect(res.data.getAllCommentsByUser).toBeDefined();
    expect(res.data.getAllCommentsByUser).toContainEqual({
      product: { _id: productId },
      text: newComment.text,
      user: newComment.user,
      show: newComment.show,
    });
    done();
  });
  it(' Should receive all comments for one product', async done => {
    const res = await operations
      .query({
        variables: {
          productId,
        },
        query: gql`
          query($productId: ID!) {
            getAllCommentsByProduct(productId: $productId) {
              items {
                text
                product {
                  _id
                }
                show
                user {
                  email
                }
              }
            }
          }
        `,
      })
      .catch(e => e);
    const receivedComments = res.data.getAllCommentsByProduct.items;
    expect(receivedComments).toBeDefined();
    expect(receivedComments).toContainEqual({
      product: { _id: productId },
      text: newComment.text,
      user: newComment.user,
      show: newComment.show,
    });
    done();
  });
  it(' Should receive all comments for one product', async done => {
    const res = await operations
      .query({
        variables: {
          productId: productWrongId,
        },
        query: gql`
          query($productId: ID!) {
            getAllCommentsByProduct(productId: $productId) {
              items {
                text
                product {
                  _id
                }
                show
                user {
                  email
                }
              }
            }
          }
        `,
      })
      .catch(e => e);
    const error = res.errors[0].message;

    expect(error).toBeDefined();
    expect(error).toBe(COMMENT_NOT_FOUND);
    done();
  });

  it(' should return empty array of comments for unexisting email ', async done => {
    const res = await operations.query({
      variables: {
        userEmail: invalidEmail,
      },
      query: gql`
        query($userEmail: String!) {
          getAllCommentsByUser(userEmail: $userEmail) {
            ... on Comment {
              text

              product {
                _id
              }
              show
              user {
                email
              }
            }
          }
        }
      `,
    });
    expect(res.data.getAllCommentsByUser).toBeDefined();
    expect(res.data.getAllCommentsByUser.length).toBe(0);
    expect(res.data.getAllCommentsByUser).toBeInstanceOf(Array);
    done();
  });

  it(' should return one comment', async done => {
    const res = await operations
      .query({
        variables: {
          id: commentId,
        },
        query: gql`
          query($id: ID!) {
            getCommentById(id: $id) {
              ... on Comment {
                text
                product {
                  _id
                }
                show
                user {
                  email
                }
              }
              ... on Error {
                message
                statusCode
              }
            }
          }
        `,
      })
      .catch(e => e);
    const receivedComment = res.data.getCommentById;
    expect(receivedComment).toBeDefined();

    expect(receivedComment).toBeDefined();
    expect(receivedComment).toEqual({
      product: { _id: productId },
      text: newComment.text,
      user: newComment.user,
      show: newComment.show,
    });
    done();
  });
  it(' should return error when find comment by wrong id', async done => {
    const res = await operations
      .query({
        variables: {
          id: commentWrongId,
        },
        query: gql`
          query($id: ID!) {
            getCommentById(id: $id) {
              ... on Comment {
                text
                product {
                  _id
                }
                show
                user {
                  email
                }
              }
              ... on Error {
                message
                statusCode
              }
            }
          }
        `,
      })
      .catch(e => e);
    const receivedComment = res.data.getCommentById;
    expect(receivedComment).toBeDefined();
    expect(receivedComment).toMatchSnapshot();
    expect(receivedComment).toBeDefined();
    expect(receivedComment).toHaveProperty('statusCode', 404);
    expect(receivedComment).toHaveProperty('message', COMMENT_NOT_FOUND);
    done();
  });
});
