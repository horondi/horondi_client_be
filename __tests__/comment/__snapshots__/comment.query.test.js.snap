// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment queries  should return error when find comment by wrong id 1`] = `
Object {
  "message": "COMMENT_NOT_FOUND",
  "statusCode": 404,
}
`;
